File: 2018/round-A/EvenDigits.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// PROBLEM
// Supervin has a unique calculator. This calculator only has a display, a plus button, and a minus button.
// Currently, the integer N is displayed on the calculator display.
// Supervin does not like odd digits, because he thinks they are "odd". Therefore, he wants to display an integer with
// only even digits in its decimal representation, using only the calculator buttons.
// Since the calculator is a bit old and the buttons are hard to press, he wants to use a minimal number of button presses.

int length; // the length of the displayed number

// This functions compares 2 long long integers and returns the minimum
long long minimum(long long a, long long b) {
    return (a > b ? b : a);
}

// This function returns 1 if the parameter is odd, 0 if it's even
int isOdd(int value) {
    return value % 2;
}

// This function writes in the target string the largest lower even number
void lowerEven(char* target, int value, int position) {
    int i;

    // Particular case: the lower even is shorter by 1 digit
    if (value == 1 && position == 0) {
        for (i = 0; i < length - 1; ++i) {
            *(target + i) = '8';
        }
        *(target + i) = '\0';
    }

    // General case: decrease by 1 the odd digit and put the right number of '8's after
    else {
        *(target + position) = (char) value + 47;
        for (i = position + 1; i < length; i++) {
            *(target + i) = '8';
        }
        *(target + i) = '\0';
    }
}

// This function writes in the target string the smallest higher even number
void higherEven(char* target, int value, int position) {
    int i;

    if (value >= 8) {

        // Particular case: if the digit to be increased is '8' or '9' and it's the leftmost digit, then the target
        // string is formed by a '2' followed by as many '0's as the length of the initial number
        if (position == 0) {
            *(target) = '2';
            for (i = 1; i <= length; ++i) {
                *(target + i) = '0';
            }
            *(target + i) = '\0';
        }

        // Particular case: if the digit to be increased is '8' or '9', then we recursively call this function on the
        // previous digit
        else {
            int precedent = (int) *(target + position - 1) - 48;
            return higherEven(target, precedent, position - 1);
        }
    }

    // General case: increase the digit by 1 (if it is odd) or 2 (if it is even) and put the correct number of '0's.
    else {
        *(target + position) = (char) value + (value % 2 ? 49 : 50);
        for (i = position + 1; i < length; ++i) {
            *(target + i) = '0';
        }
        *(target + i) = '\0';
    }
}

int main() {
    int t;              // Number of test cases
    scanf("%d", &t);    // Read the number of test cases from stdin

    for (int counter = 1; counter <= t; ++counter) {

        char* number = (char*) malloc(20);  // The biggest possible number is 10^16 that has 17 digits, so 20 bytes is
        scanf("%s", number);                // more than actually needed

        length = strlen(number);
        char* x = (char*) malloc(length + 1);   // x is the string that will contain the lower even number
        char* y = (char*) malloc(length + 2);   // y is the string that will contain the higher even number

        char digit = *number;
        for (int i = 0; digit != '\0'; i++) {
            int value = (int) digit - 48;

            // when an odd digit is found the lower and higher even numbers are generated
            if (isOdd(value)) {
                lowerEven(x, value, i);
                higherEven(y, value, i);
                break;
            }

            // the original number is transcribed in x and y as long as the digits are even
            else {
                *(x + i) = digit;
                *(y + i) = digit;
            }

            digit = *(number + i + 1);
        }

        // conversions from string to long long
        long long numberValue = atoll(number);
        long long low = atoll(x);
        long long high = atoll(y);

        // compute the closer even number
        long long closer = minimum(llabs(numberValue - low), llabs(numberValue - high));
        printf("Case #%d: %lld\n", counter, closer);
    }
}

File: 2018/round-A/even_digits.py
''' PROBLEM
Supervin has a unique calculator. This calculator only has a display, a plus button, and a minus button. Currently, the integer N is displayed on the calculator display.
Pressing the plus button increases the current number displayed on the calculator display by 1. Similarly, pressing the minus button decreases the current number displayed on the calculator display by 1. 
The calculator does not display any leading zeros. For example, if 100 is displayed on the calculator display, pressing the minus button once will cause the calculator to display 99.
Supervin does not like odd digits, because he thinks they are "odd". Therefore, he wants to display an integer with only even digits in its decimal representation, using only the calculator buttons. 
Since the calculator is a bit old and the buttons are hard to press, he wants to use a minimal number of button presses.
Please help Supervin to determine the minimum number of button presses to make the calculator display an integer with no odd digits.
'''

import sys

# digit is odd when called
def bounds(digit, digits_left):
	
	# complete the number with left digits. 
	# higher is wrong if digit is 9, anyway its value is ignored in that case 
	lower = str(digit - 1)
	higher = str(digit + 1)
	for _ in range(digits_left):
		lower = lower + '8'
		higher = higher + '0'
	return lower, higher

def compute_min_presses(N):
	result = ''
	digits_left = len(N)
	for character in N:
		digits_left -= 1
		digit = int(character)

		# if the digit is odd, then compute the higher and lower even
		if digit % 2 == 1:
			lower_part, higher_part = bounds(digit, digits_left)
			lower_even = int(result + lower_part)
			lower_diff = int(N) - lower_even

			# if the odd digit is 9, then the lower even is closer than the higher for sure
			if digit == 9:
				return lower_diff

			higher_even = int(result + higher_part)
			higher_diff = higher_even - int(N)

			# the minimum button presses reaches the closest even to N
			return min(lower_diff, higher_diff)

		# if this digit is even transcribe the digit to result
		else:
			result = result + character

	# if every digit is even, no button presses should be performed
	return 0
		

# read the numebr of test cases
T = int(sys.stdin.readline())
result = ''
for i in range(T):

	# read the value of the test case
	N = sys.stdin.readline().rstrip('\n')

	# compute minimum button presses
	x = i + 1
	y = compute_min_presses(N)
	result += f'Case #{x}: {y}\n'

print(result)
